src/devices/video/upd7220.cpp

```diff
@@ -999,58 +985,155 @@ void upd7220_device::draw_rectangle(int x, int y)
 //  draw_char -
 //-------------------------------------------------
 
-void upd7220_device::draw_char(int x, int y)
+void upd7220_device::draw_char()
 {
-	int isize,psize;
-	uint16_t tile_data;
+	const int8_t dir_change[2][4] = {
+		{2, 2, -2, -2},
+		{1, 3, -3, -1}
+	};
+	const uint8_t type = (m_figs.m_figure_type & 0x10) >> 4;
 
-	LOG("uPD7220 char check: %d %d %02x %08x %d %d %02x\n",x,y,m_figs.m_dir,m_ead,m_figs.m_d,m_figs.m_dc,m_figs.m_figure_type);
+	uint16_t pattern;
 
-	/* QX10 may require upper bits for psize, VT240 requires the opposite */
-	isize = m_figs.m_d;
-	psize = m_figs.m_dc + 1;
+	LOG("uPD7220 char check: %08x %04x %02x %02x %d %d %02x\n",m_ead,m_mask,m_bitmap_mod,m_figs.m_dir,m_figs.m_d,m_figs.m_dc,m_figs.m_figure_type);
 
-	for(int pi = 0; pi < psize; pi++)
+	for (int i = 0, di = 0; i < (m_figs.m_dc + 1); ++i)
 	{
-		tile_data = (m_ra[15-(pi & 7)] << 8) | m_ra[15-(pi & 7)];
-		for(int pz = 0; pz <= m_gchr; pz++)
+		m_pattern = (m_ra[15 - (i & 7)] << 8) | m_ra[15 - (i & 7)];
+		for (int zdc = 0; zdc <= m_gchr; ++zdc, ++di)
 		{
-			int ii = 0, curpixel = 0;
-			if(pi & 1)
+			for (int j = 0; j < m_figs.m_d; ++j)
 			{
-				ii = isize - 1;
-				curpixel = (isize * (m_gchr + 1)) - 1;
-			}
-			while(pi & 1 ? ii >= 0 : ii < isize)
-			{
-				for(int iz = 0; iz <= m_gchr; iz++)
+				if (di % 2) {
+					pattern = get_pattern(15-(j % 0xf));
+				} else {
+					pattern = get_pattern((j % 0xf));
+				}
+				for (int zd = 0; zd <= m_gchr; ++zd)
 				{
-					draw_pixel(x + (curpixel * x_dir[m_figs.m_dir]), y + (curpixel * y_dir[m_figs.m_dir]), ii, tile_data);
-					if(pi & 1)
-						curpixel--;
-					else
-						curpixel++;
+					write_vram(0, m_bitmap_mod, pattern, m_mask);
+
+					if (!(j == (m_figs.m_d - 1) && zd == m_gchr))
```

Comment:
This expression would be clearer if you used de Morgan’s theorem to simplify it.